### PR TITLE
Merge two config maps into one that defines control switches and user defined injections

### PR DESCRIPTION
**Pull request by MichalGuzieniuk**
This pull requests aims to merge User Defined Injections (UDI) ConfigMap into the control switches ConfigMap.

In result, there will be only one map to provided by the user. From the code perspective I refactored code by moving out from webhook.go all functions related to UDI into separate package called custominjections. This way, there is a clear information about package responsibility.

Fixes https://github.com/k8snetworkplumbingwg/network-resources-injector/issues/61

### CODE REVIEW

**Review by Eoghan1232 — APPROVED** [review]

**Comment by Eoghan1232** [comment]
@zshi-redhat Could you review?

File: pkg/custominjections/custominjections.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// CreateUserInjectionsStructure returns empty UserDefinedInjections structure
+func CreateUserInjectionsStructure() *UserDefinedInjections {
+	var userDefinedInjects = UserDefinedInjections{Patchs: make(map[string]types.JsonPatchOperation)}
+	return &userDefinedInjects
+}
+
+// SetCustomizedInjections sets additional injections to be applied in Pod spec
+func (userDefinedInjects *UserDefinedInjections) SetCustomizedInjections(injectionsCm *corev1.ConfigMap) {
```
Comment (zshi-redhat) [inline pkg/custominjections/custominjections.go:32]: Please use the consistent interface name as in line 95.
Either change `userDefinedInjects` to `injections` in this line or change `injections` to `userDefinedInjects` in line 95.

File: pkg/custominjections/custominjections.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
+package custominjections
```
Comment (zshi-redhat) [inline pkg/custominjections/custominjections.go:1]: I noticed there are some remainings when I updated the `custom injection` term to `user defined injection`. Would you mind updating this PR to use `user defined injection`? Ideally we would like to replace custom to user-defined.

**Review by zshi-redhat — COMMENTED** [review]
@MichalGuzieniuk Thanks for the PR and sorry for the late review!

File: pkg/custominjections/custominjections.go (hunk 1 above)
Comment (MichalGuzieniuk) [reply pkg/custominjections/custominjections.go]: Done.

File: pkg/custominjections/custominjections.go (hunk 2 above)
Comment (MichalGuzieniuk) [reply pkg/custominjections/custominjections.go]: Done.

File: pkg/userdefinedinjections/userdefinedinjections.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// CreateUserInjectionsStructure returns empty UserDefinedInjections structure
+func CreateUserInjectionsStructure() *UserDefinedInjections {
+	var userDefinedInjects = UserDefinedInjections{Patchs: make(map[string]types.JsonPatchOperation)}
+	return &userDefinedInjects
+}
+
+// SetCustomizedInjections sets additional injections to be applied in Pod spec
+func (userDefinedInjects *UserDefinedInjections) SetCustomizedInjections(injectionsCm *corev1.ConfigMap) {
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections.go:32]: SetCustomizedInjections -> SetUserDefinedInjections

File: pkg/userdefinedinjections/userdefinedinjections.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			userDefinedInjects.Patchs = make(map[string]types.JsonPatchOperation)
+		}
+	} else {
+		glog.Warningf("Map does not contains [%s]. Clear old entries", types.ConfigMapMainFileKey)
+		userDefinedInjects.Patchs = make(map[string]types.JsonPatchOperation)
+	}
+}
+
+// CreateCustomizedPatch creates customized patch for the specified POD
+func (userDefinedInjects *UserDefinedInjections) CreateCustomizedPatch(pod corev1.Pod) ([]types.JsonPatchOperation, error) {
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections.go:95]: CreateCustomizedPatch -> CreateUserDefinedPatch

File: pkg/userdefinedinjections/userdefinedinjections_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+package userdefinedinjections
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestCustomInjections(t *testing.T) {
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections_suite_test.go:24]: TestUserDefinedInjections

File: pkg/userdefinedinjections/userdefinedinjections_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "CustomInjections Suite")
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections_suite_test.go:26]: CustomInjections -> UserDefinedInjections

File: pkg/userdefinedinjections/userdefinedinjections_test.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/k8snetworkplumbingwg/network-resources-injector/pkg/types"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("CustomInjections", func() {
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections_test.go:28]: CustomInjections > UserDefinedInjections

File: pkg/userdefinedinjections/userdefinedinjections_test.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	DescribeTable("Create user-defined patchs",
+
+		func(pod corev1.Pod, userDefinedInjectPatchs map[string]types.JsonPatchOperation, out []types.JsonPatchOperation) {
+			userDefinedInjects := CreateUserInjectionsStructure()
+			userDefinedInjects.Patchs = userDefinedInjectPatchs
+			appliedPatchs, _ := userDefinedInjects.CreateCustomizedPatch(pod)
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections_test.go:34]: CreateCustomizedPatch > CreateUserDefinedPatch

File: pkg/userdefinedinjections/userdefinedinjections_test.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+			},
+			nil,
+		),
+	)
+
+	DescribeTable("Setting user-defined injections",
+		func(in *corev1.ConfigMap, existing map[string]types.JsonPatchOperation, out map[string]types.JsonPatchOperation) {
+			userDefinedInjects := CreateUserInjectionsStructure()
+			userDefinedInjects.Patchs = existing
+			userDefinedInjects.SetCustomizedInjections(in)
```
Comment (zshi-redhat) [inline pkg/userdefinedinjections/userdefinedinjections_test.go:103]: SetCustomizedInjections -> SetUserDefinedInjections

File: pkg/webhook/webhook.go [hunk 10]
```diff
@@ -735,12 +709,12 @@ func appendAddAnnotPatch(patch []jsonPatchOperation, pod corev1.Pod, userDefined
 	return patch
 }
 
-func appendCustomizedPatch(patch []jsonPatchOperation, pod corev1.Pod, userDefinedPatch []jsonPatchOperation) []jsonPatchOperation {
+func appendCustomizedPatch(patch []types.JsonPatchOperation, pod corev1.Pod, userDefinedPatch []types.JsonPatchOperation) []types.JsonPatchOperation {
```
Comment (zshi-redhat) [inline pkg/webhook/webhook.go:712]: appendCustomizedPatch > appendUserDefinedPatch

File: pkg/webhook/webhook.go [hunk 11]
```diff
@@ -788,7 +762,7 @@ func MutateHandler(w http.ResponseWriter, req *http.Request) {
 	}
 	glog.Infof("AdmissionReview request received for pod %s/%s", pod.ObjectMeta.Namespace, pod.ObjectMeta.Name)
 
-	userDefinedPatch, err := createCustomizedPatch(pod)
+	userDefinedPatch, err := userDefinedInjections.CreateCustomizedPatch(pod)
```
Comment (zshi-redhat) [inline pkg/webhook/webhook.go:765]: CreateCustomizedPatch > CreateUserDefinedPatch

File: cmd/webhook/main.go [hunk 12]
```diff
@@ -207,16 +210,7 @@ func main() {
...
-			cm, err = clientset.CoreV1().ConfigMaps(namespace).Get(
-				context.Background(), userDefinedInjectionConfigMap, metav1.GetOptions{})
-			if err != nil {
-				if !errors.IsNotFound(err) {
-					glog.Warningf("Failed to get configmap for user-defined injections: %v", err)
-					continue
-				}
-			}
-			webhook.SetCustomizedInjections(cm)
+			userInjections.SetCustomizedInjections(cm)
```
Comment (zshi-redhat) [inline cmd/webhook/main.go:213]: SetCustomizedInjections -> SetUserDefinedInjections

File: README.md [hunk 13]
```diff
@@ -263,10 +263,21 @@ Below is an example of user defined injection ConfigMap:
...
 kind: ConfigMap
 metadata:
-  name: nri-user-defined-injections
+  name: nri-control-switches
   namespace: kube-system
 data:
-  feature.pod.kubernetes.io_sriov-network: '{"op": "add", "path": "/metadata/annotations", "value": {"k8s.v1.cni.cncf.io/networks": "sriov-net-attach-def"}}'
+  config.json: |
+    {
+      "customInjections": {
```
Comment (zshi-redhat) [inline README.md:271]: customInjections > user-defined-injections

**Review by zshi-redhat — COMMENTED** [review]
A few more places to replace custom with user-defined. see comments.

File: pkg/userdefinedinjections/userdefinedinjections.go (hunk 3 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections.go]: Done.

I should do a training from search tool ;) - how could I missed that. Sorry for double work on review.

Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections.go]: Done

File: pkg/userdefinedinjections/userdefinedinjections.go (hunk 4 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections.go]: Done.

File: pkg/userdefinedinjections/userdefinedinjections_suite_test.go (hunk 5 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections_suite_test.go]: Done.

File: pkg/userdefinedinjections/userdefinedinjections_suite_test.go (hunk 6 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections_suite_test.go]: Done.

File: README.md (hunk 13 above)
Comment (MichalGuzieniuk) [reply README.md]: Done.

File: cmd/webhook/main.go (hunk 12 above)
Comment (MichalGuzieniuk) [reply cmd/webhook/main.go]: Done.

File: pkg/webhook/webhook.go (hunk 11 above)
Comment (MichalGuzieniuk) [reply pkg/webhook/webhook.go]: Done.

File: pkg/userdefinedinjections/userdefinedinjections_test.go (hunk 7 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections_test.go]: Done.

File: pkg/userdefinedinjections/userdefinedinjections_test.go (hunk 9 above)
Comment (MichalGuzieniuk) [reply pkg/userdefinedinjections/userdefinedinjections_test.go]: Done.

File: pkg/webhook/webhook.go (hunk 10 above)
Comment (MichalGuzieniuk) [reply pkg/webhook/webhook.go]: Done.

**Comment by zshi-redhat** [comment]
In the `cmd/webhook/main.go`, there is this piece of code:

```
202                 case <-time.After(30 * time.Second):
203                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
204                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
205                         if err != nil {
206                                 glog.Warningf("Error within control switches map %s", err.Error())
207                                 if !errors.IsNotFound(err) {
208                                         glog.Info("Map with runtime configuration not available using default / CLI settings")
209                                 }
210                         }
211                         // has to be called each time because we need to restore default config when map is empty
212                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
213                         userInjections.SetUserDefinedInjections(cm)
214                 }
```

Could we use the following:

```
                        cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
                                context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
                        if err != nil {
                                if !errors.IsNotFound(err) {
                                        glog.Warningf("Error getting control switches configmap %s", err.Error())
                                }
                                continue
                        }   
                        // has to be called each time because we need to restore default config when map is empty
                        controlSwitches.ProcessControlSwitchesConfigMap(cm)
                        userInjections.SetUserDefinedInjections(cm)
                }
```

So that it won't print a Warningf message every 30 seconds and the processing of controlSwitch configmap would be skipped, when there is no controlSwitch configmap defined.

Also could you merge the three commits into one?

Thank you!

**Comment by MichalGuzieniuk** [comment]
> In the `cmd/webhook/main.go`, there is this piece of code:
> 
> ```
> 202                 case <-time.After(30 * time.Second):
> 203                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
> 204                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
> 205                         if err != nil {
> 206                                 glog.Warningf("Error within control switches map %s", err.Error())
> 207                                 if !errors.IsNotFound(err) {
> 208                                         glog.Info("Map with runtime configuration not available using default / CLI settings")
> 209                                 }
> 210                         }
> 211                         // has to be called each time because we need to restore default config when map is empty
> 212                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
> 213                         userInjections.SetUserDefinedInjections(cm)
> 214                 }
> ```
> 
> Could we use the following:
> 
> ```
>                 case <-time.After(30 * time.Second):
>                         cm, err := clientset.CoreV1().ConfigMaps(namespace).Get(
>                                 context.Background(), controlSwitchesConfigMap, metav1.GetOptions{})
>                         if err != nil {
>                                 if !errors.IsNotFound(err) {
>                                         glog.Warningf("Error getting control switches configmap %s", err.Error())
>                                 }
>                                 continue
>                         }   
>                         // has to be called each time because we need to restore default config when map is empty
>                         controlSwitches.ProcessControlSwitchesConfigMap(cm)
>                         userInjections.SetUserDefinedInjections(cm)
>                 }
> ```
> 
> So that it won't print a Warningf message every 30 seconds and the processing of controlSwitch configmap would be skipped, when there is no controlSwitch configmap defined.
> 
> Also could you merge the three commits into one?
> 
> Thank you!

I have removed the log that was polluting the events log. But please notice that I did not add **continue**'in case where there  error. According to what we agree within issue description, in case when there is no config map or it is invalid - NRI should restore default values.

I squashed also all commits into one, but could you explain me why I should do it? I thought that it is possible to squash commits when merge is done from my fork to npwg repo directly by gihub.

**Comment by zshi-redhat** [comment]
> I have removed the log that was polluting the events log. But please notice that I did not add **continue**'in case where there error. According to what we agree within issue description, in case when there is no config map or it is invalid - NRI should restore default values.

In case there is no configmap - yes, I think it should re-initialize the settings.
In case the configmap is invalid - yes, NRI should restore default values (how do we define invalid, does it mean the data of configmap is not in right format, but configmap itself is retrievable by NRI?).

I think there is a third condition:
In case the NRI fails to get the configmap due to a temporary api unaccessibility, the settings will be re-initialized until the next api call success. I assume we would want to avoid the temporary change on the settings.

> 
> I squashed also all commits into one, but could you explain me why I should do it? I thought that it is possible to squash commits when merge is done from my fork to npwg repo directly by gihub.

Yes, you're right, the squash can be done in merge action, I will do that next time.

**Comment by MichalGuzieniuk** [comment]
> In case there is no configmap - yes, I think it should re-initialize the settings. In case the configmap is invalid - yes, NRI should restore default values (how do we define invalid, does it mean the data of configmap is not in right format, but configmap itself is retrievable by NRI?).
> 
> I think there is a third condition: In case the NRI fails to get the configmap due to a temporary api unaccessibility, the settings will be re-initialized until the next api call success. I assume we would want to avoid the temporary change on the settings.
> 


Done.

**Review by zshi-redhat — APPROVED** [review]

**Comment by zshi-redhat** [comment]
@Eoghan1232 would you want to take another look since there has been some update since your last approve?

**Comment by Eoghan1232** [comment]
> @Eoghan1232 would you want to take another look since there has been some update since your last approve?

Yes, I will review once more.

**Pull request merged**